src/hotspot/share/gc/g1/g1OopClosures.hpp

```diff
@@ -68,6 +66,16 @@ class G1ScanCardClosure : public G1ScanClosureBase {
                     size_t& heap_roots_found) :
     G1ScanClosureBase(g1h, pss), _heap_roots_found(heap_roots_found) { }
 
+  // Because this closure is applied on pointers residing outside the
+  // collection set, we shouldn't do discovery, which is why this closure has
+  // its reference-processor being null.
+  // Strictly speaking, one can use the same iteration mode from the superclass
+  // BasicOopIterateClosure, and the null reference-processor will treat fields
+  // as strong references anyway, equivalent to DO_FIELDS. Here we override the
+  // iteration mode to skip the known null-check in
+  // InstanceRefKlass::try_discover.
+  virtual ReferenceIterationMode reference_iteration_mode() { return DO_FIELDS; }
```

Comment:
I think the comment is okay, but maybe should put more emphasis on that specifying this iteration mode allows _the compiler_ to (more easily) compile out the check in `try_discover`.

Also since this isn't the only place this (and similar) optimizations are made using this getter, I kind of agree with @lgxbslgx that the impact of the selection should probably be documented at the enum level (the move of the override into the leaf class here is fine with me).

Just something like "// Selection of one or the other option helps the compiler to remove unnecessary, known beforehand behavior for a given oop closure at compile time" seems sufficient to me there instead of this fairly long comment.
The exact changes can be easily gathered by searching the usages of the enum.

Reply:
I dropped the comment. This optimization, afaics, is only used/implemented in G1. Moving the comment to the enum level can be confusing, since its impact might not be visible (or even positive) for other collectors. (I inspected the generated asm of `g1Remset` with and without this optimization and they are indeed different -- the one *without* the optimization is shorter (in terms of LOC), but it's unclear which one is faster.)

Reply:
I can see multiple references to the `DO_FIELDS` value in Hotspot code in that getter in several GCs, but only grepped for it without looking into detail about the reasons.